lib/TPP/Transforms/LoopShuffle.cpp

```diff
@@ -0,0 +1,109 @@
+//===- LoopShuffle.cpp -----------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file shuffles parallel loop based on user input
+//
+//===----------------------------------------------------------------------===//
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include "llvm/Support/Debug.h"
+#include <list>
+
+#define DEBUG_TYPE "loop-shuffle"
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DECL_LOOPSHUFFLEPASS
+#define GEN_PASS_DEF_LOOPSHUFFLEPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::scf;
+using namespace std;
+
+namespace mlir {
+namespace tpp {
+
+static LogicalResult loopShuffle(scf::ForallOp op,
+                                 ArrayRef<unsigned> shuffleOrder) {
+  OpBuilder b(op);
+  IRRewriter rewriter(b.getContext());
+  if (shuffleOrder.size() != op.getInductionVars().size())
+    return rewriter.notifyMatchFailure(op, "Number of indices incorrect");
+  for (size_t i = 0; i < op.getInductionVars().size(); i++) {
+    bool match = false;
+    for (size_t j = 0; j < shuffleOrder.size(); j++)
```

Comment:
Easier to sort the list and check if `shuffleOrder[i] = i`.

Reply:
We'd have to make a copy of shuffle order and then sort it.